src/xds/metrics.rs

```diff
@@ -58,3 +85,18 @@ impl Recorder<ConnectionTerminationReason, u64> for Metrics {
             .inc_by(count);
     }
 }
+
+impl Recorder<DeltaDiscoveryResponse, ()> for Metrics {
```

Comment:
```suggestion
impl Recorder<DeltaDiscoveryResponse, u64> for Metrics {
```

The intent here is this is ~always a u64. I am not even sure why we make it generic, I think we maybe used to have some float usage.

Then `IncrementRecorder` auto-implements and gives `.increment()` btw

Reply:
But which metric would this increment? The number of bytes or the resource type? My thought process is to have one `impl` to handle all metrics for a single request. That way, we aren't indexing the metric by the metric's type, which will cause collisions if we have two metrics of type, say, `u64`.

Reply:
The type of the metric isn't u64, its the type of the value.

But it is maybe a bit weird since someone could do `record(10)`, which is nonsense -- you can only really meaningfull `increment()` it

Reply:
I think we have different understandings of the code.

My impression of the current implementation of `Recorder` and `IncrementRecorder` is that they deal with only one metric at a time.
For example, `fn record(&self, reason: &ConnectionTerminationReason, count: u64)` updates the `Metrics.connection_termination` entry corresponding to `reason` by `count`.
But if I used `u64` instead of `()` for `Recorder`, then the signature for `record` would be `record(&self, response: &DeltaDiscoveryResponse, x: u64)`. This `record` implementation would update two metrics: `Metrics.message_types`, and `Metrics.total_message_size`. In this case. what would `x` be? Is it the number of bytes? The number of resources of a certain type? If so, of which type? Regardless, this information is already in `response`. Even in the `increment` case (which is really just `record(response, 1)`), I still find the second argument to be confusing. I just don't see a need for the second argument, hence the `()` type.

I could create multiple `record` functions, one for each metric, but I feel like we run into the same problem.
If I key `total_message_size` by `TypeUrl` and have the function signature be `record(&self, response: &TypeUrl, count: u64)`, it isn't clear which metric would be updated by `count`. Is it `message_types` or `total_message_size`? Regardless, having one `record` for each `DeltaDiscoveryResponse` seems better to me because we only have to make one call per response instead of however many metrics we have.

We could also just ignore the second argument of `record` (`fn record(response: &DeltaDiscoveryResponse, _: u64)`) and then just do `increment(response)`, but that feels like the same thing with more steps.

I realize that this probably doesn't reflect what you're thinking and I probably don't really understand what you're saying, but I hope my thought process is more clear now.